content/exchange/artifacts/FreeBSD.Sys.Utx.yaml

```diff
@@ -0,0 +1,76 @@
+
+type: CLIENT
+
+parameters:
+  - name: utxGlobs
+    default: /var/log/utx.log*
+    description: |
+      glob for covering the files that should be parsed; default covers the usual location of the utx file on FreeBSD
+  - name: userRegex
+    type: regex
+    default: "."
+    description: |
+      Regex for filtering the users, showing those you are interested in 
+  - name: DateAfter
+    type: timestamp
+    description: |
+      timestamp used for filtering the login time
+  - name: DateBefore
+    type: timestamp
+    description: |
+      timestamp used for filtering the login time
+
+sources:
+  - precondition:
+      SELECT OS From info() where OS = 'freebsd'
+    query: |
+      -- timestamps given by the system command "last" are in local time, so we tell Velociraptor to handle them accordingly when converting with the VQL function "timestamp()"
+      LET PARSE_TZ <= "local"
+      
+      -- time test function (taken from Windows.NTFS.MFT)
+      LET time_test(stamp) =
+            if(condition= DateBefore AND DateAfter,
+                then= stamp < DateBefore AND stamp > DateAfter,
+                else=
+            if(condition=DateBefore,
+                then= stamp < DateBefore,
+                else=
+            if(condition= DateAfter,
+                then= stamp > DateAfter,
+                else= True
+            )))
+      
+      -- expand the glob and get all files that are matched by it
+      LET Files = SELECT OSPath FROM glob(globs=split(string=utxGlobs, sep=","))
+      
+      -- for each targeted file, parse the data out of it using system built-in command 
+      LET UtxParsing = SELECT * FROM foreach(row=Files,
+      query={
+        -- TODO: try the command while setting env variable "TZ" to UTC and check, if the timestamps are changed accordingly
+        SELECT * 
+        FROM execve(argv=["last", "-wy", "--libxo=json", "-f", OSPath])
+        WHERE log(message="stderr: " + Stderr) 
+      })
+      
+      -- parse the output from each file as json
+      LET UtxContent = SELECT parse_json(data=Stdout) as Content FROM UtxParsing
+      
+      -- pull out the interesting information from the nested json content
+      LET FormatedContent = SELECT * FROM foreach(row=UtxContent,
+      query={
+        -- the if-blocks are added because some Symbols are not present in each row, creating ugly errors in the flow if naively selected
+        SELECT user, timestamp(string=`login-time`) AS login_time, if(condition={SELECT * FROM scope() WHERE `logout-time`},then=`logout-time`,else="") AS logout_time, if(condition={SELECT * FROM scope() WHERE `session-length`},then=`session-length`,else="") AS session_length, if(condition={SELECT * FROM scope() WHERE tty},then=tty,else="") AS tty, if(condition={SELECT * FROM scope() WHERE `from`},then=`from`,else="") AS remote, Content.`last-information`.utxdb AS utx_log FROM foreach(row=Content.`last-information`.last)
```

Comment:
I'll have to test that and will report back, if I can simplify that part of the query.
I created the construct to solve the "problem" that some entries in `Content."last-information".last` do not have an element "logout-time". This would create errors, which showed the flow in the UI as having errored (which I thought might confuse people using the artifact).

Reply:
not a problem - usually you can suppress the symbol not found errors using something like

```sql
LET s = scope()

SELECT ... s.logout_time AS logout_time ....
FROM ....
```

This works because s is the scope dict and s.logout_time means to pull out the logout_time member in that dict- so if it is not there the default is just NULL not a symbol not found error.

Also there is no need to surround logout_time in backticks because its name does not contain any VQL operators (like for e.g. space or dots)

Reply:
The two statements seem not to be the same. I changed the query as you suggested, but now I get following errors, which mark the flow as having errored:
![image](https://github.com/user-attachments/assets/54d7fcc2-1234-41da-a2c1-21898e2399c7)

I rather prefer having the flow not show as errored, as this might confuse an analyst using it.

Reply:
with your last suggestion, it works and the query looks a lot nicer.
Thanks :)